### DISCUSSION TITLE
AutowireViewModel is not working as expected in migrated MAUI app

### DISCUSSION THREAD

**Discussion by AngeloLazzari** [PrismLibrary/Prism#3258] [.NET MAUI]
Hello, in our old xamarin app we were using the AutowireViewModel=true and now we put the new AutowireViewModel=Automatic property and we left all the rest of the MVVM code like it was.

As you see in the image we set the internal viewmodel variable to the "BindingContext as ..." clause but the problem we have is that once the line is executed the variable is still null.

![mvvm](https://github.com/user-attachments/assets/ce47c12d-352c-41a4-bb46-1e2e6ef22fe9)

How can we check that the AutowireViewModel met the correct viewmodel? 

Thank you very much!

**Comment by AngeloLazzari** [comment]
and here i have the watches ar runtime... 
![erro mvvm](https://github.com/user-attachments/assets/91f4d321-aada-4838-9184-ba999c129ff8)

I spent time to find a solution but I didn't find the cause...
Thankyou

**Comment by brianlagunas** [comment]
The behavior of the ViewModelLocator has changed from the Xamarin version. You are most likely checking the value too soon since the VM is not set during construction of the View but rather after. This means the CTOR is too early in the lifecycle. One thing you could try is to set the ViewModelLocator to `Forced` instead of `Automatic`. I believe that will revert to the prior behavior of the VML and checking in the CTOR might work in that case.

**Reply by dansiegel** [reply]
To expand on this. There are a number of reasons why creating the ViewModel immediately is bad in .NET MAUI. As a result the default behavior is to create the ViewModel after the ctor instead of as part of the InitializeComponent.

I would recommend that you either just do:

```cs
private MyViewModel ViewModel => (MyViewModel)BindingContext;
```

Or you can override the OnBindingContextChanged method. Keep in mind that this also means that the ViewModel may go null at the end of the View's lifecycle.

**Reply by AngeloLazzari** [reply]
Thanks Guys, I will try now to apply your advises. You are a great team! thank you for your help!!!!

**Comment by AngeloLazzari** [comment]
Hi, I am back again, sorry for the late reply but I needed to do crazy test to check what is happening.

So now I put the autoowire to Force and the private MyViewModel ViewModel => (MyViewModel)BindingContext;

But now I have a different error: page to remove must be containd on this Navigation Page.. (see picture)
![image](https://github.com/user-attachments/assets/d87ec2dc-44b7-495f-ab88-972bc6b8f103)

I searched for the error and I met this https://github.com/PrismLibrary/Prism/issues/3153

Cheking my code I have ( I think...) all correclty set... I mean, the registration of the navigation:
![image](https://github.com/user-attachments/assets/df2517f7-6896-4d5a-866a-98ff8b37f07a)

the navigation call:
![image](https://github.com/user-attachments/assets/39fb6d60-209e-461e-b9c7-479950cef862)

What I am doing wrong? thank you very much

**Reply by dansiegel** [reply]
> What I am doing wrong?

To be honest I'd be starting with looking at why you're trying to force the initialization of the ViewModel earlier than it's recommended. I may have put he capability there to shut people up but if you require it, IMO it's really likely that it's a sign of other problems in your code.

> But now I have a different error: page to remove must be containd on this Navigation Page.

Without a sample that reproduces the error it's hard to say exactly what the issue could be.

**Reply by AngeloLazzari** [reply]
Thanks Dan for your reply, we are migrating this app from xamarin and all is working like a charm in Xamarin, the problem is is MAUI.
I will check the possibility to move the ViewModel initialization in the right place... 

Is this document and video still valid for MAUI? https://github.com/PrismLibrary/Prism-Samples-Forms/blob/master/13-ViewModelInitialization/ReadMe.md

thank you very much

**Comment by AngeloLazzari** [comment]
Hello! I read all the documentation about MAUI, I saw all you video on migrating the app, and how to work with Prism correctly, we checked out the Prism MAUI samples,... but no way to solve the error we have.... 

So, please, can you give a little help to me? Tell me if we are doing well?

Here is the logic we process: 

The app starts with a:
**await navigationService.NavigateAsync("/SpashScreen");**

in the MauiProgram.cs file correctly done from the .createWindow method... 

PD: we read about OnAppStart but no way to get available in code ... https://prismlibrary.github.io/docs/maui/appbuilder.html#onappstart

SpashScreen shows correctly telling that we are connecting to the server to get & check same data..... the busy cursor in looping in the screen.... so ok til here....

SpashScreen has a SpashScreenViewModel , and in the CTOR we call a method that, after some checks calls a:
**await _navigation.NavigateAsync("/ViewA");**

ViewA is a flyoutpage that has a ViewAViewModel where in the CTOR we have another navigation to:

**await _navigation.NavigateAsync("NavigationPage/ViewB");**

and here we receive the error of "page to remove must be contained on this navigation page" error.... 

Note that the only visible page we see in the simulator is the first on SpashScreen, no more page is displayed .... 

If we call the ViewC without the NavigationPage we don't have any error but the viewC is not showing... the SpashScreen is still visible.... 

Note that we have only the Prism.DryIoc.Maui package installed.

Can you give a little help or advise?

Angelo Lazzarimobile: +34 683 137 087mail: [lazzari.angelo@gmail.com](mailto:lazzari.angelo@gmail.com)


------------------------------------------------------------------------------------------------------------------------
Verify the correspondence of the addressee; otherwise, notify that to the sender and, conscious of the responsibility for the undue use, destroy the message and its copies.
------------------------------------------------------------------------------------------------------------------------

**Reply by AngeloLazzari** [reply]
@dansiegel or @brianlagunas I know your time is really really preciuos but I really need a hand on this, how can I contract a revision of the my code from your side? thank you very much!
